torch/testing/_internal/common_methods_invocations.py

```diff
@@ -704,6 +704,37 @@ def sample_inputs_index_fill(op_info, device, dtype, requires_grad):
             samples.append(SampleInput((tensor, d, idx_nonctg, fill_val)))
     return samples
 
+def sample_inputs_index_copy(op_info, device, dtype, requires_grad):
+    samples = []
+    t = make_tensor((S, S, S), device, dtype,
+                    low=None, high=None,
+                    requires_grad=requires_grad)
+    s = make_tensor((S, S, S), device, dtype,
+                    low=None, high=None,
+                    requires_grad=requires_grad)
+    # non-contiguous input
+    t01 = t.transpose(0, 1)
+    t02 = t.transpose(0, 2)
+    t12 = t.transpose(1, 2)
+    # non-contiguous input
+    s01 = t.transpose(0, 1)
+    s02 = t.transpose(0, 2)
+    s12 = t.transpose(1, 2)
+
+    # idx is a permutation of 0...S-1 for this function to be deterministic
+    idx = torch.randperm(S, dtype=torch.int64, device=device)
+    # non-contiguous index
+    idx_nonctg = torch.empty_strided((S,), (2,), device=device, dtype=torch.int64)
+    idx_nonctg.copy_(idx)
+    for d in range(t.dim()):
+        for tensor in [t, t01, t02, t12]:
+            for source in [s, s01, s02, s12]:
+                samples.append(SampleInput((tensor, d, idx, source)))
+                # Backwards is implemented with `index_select`, which does not support negative indices at the moment
+                # samples.append(SampleInput((tensor, d, -idx - 1, source)))
```

Comment:
The forward of this function also did not used to support negative indices, right? We probably don't want to change that behavior (and should add a test for it). For example, 

```
torch.index_copy(a, 0, torch.tensor((0, -1)), b)
```

used to throw 

```
RuntimeError: invalid argument 3: out of range at /pytorch/aten/src/TH/generic/THTensor.cpp:245
```

which is an odd error message that should be improved, but an error for negative indices.

Reply:
@ngimel Notes that a follow-up PR might enable negative indices, but the issue with index_select needs to be dealt with.

Reply:
I removed now the support for negative indices following your advice